platform/MpuXnLock.h

```diff
@@ -0,0 +1,71 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_MPUXNLOCK_H
+#define MBED_MPUXNLOCK_H
+
+#include "platform/mbed_mpu_mgmt.h"
+
+namespace mbed {
+
+/** \addtogroup platform */
+/** @{*/
+
+/** RAII object for disabling, then restoring RAM execute never mode
+  * Usage:
+  * @code
+  *
+  * void f() {
+  *     // some code here
+  *     {
+  *         MpuXnLock xn;
+  *         // Code in this block is allowed to call functions in RAM
+  *     }
+  *     // Execution from RAM is no longer allowed
+  * }
+  * @endcode
+  */
+class MpuXnLock {
```

Comment:
@kjbracey-arm when preparing to write docs for this class I noticed that the only lock classes Mbed OS has documented follow the previous naming convention - [DeepSleepLock](https://os.mbed.com/docs/v5.10/apis/deepsleeplock.html) and [CriticalSectionLock](https://os.mbed.com/docs/v5.10/apis/criticalsectionlock.html). Although ScopedLock and ScopedMutexLock are in the code, I don't see official documentation for them.

Should I change this back to the previous naming to match the convention of the documented code? @bulislaw @pan- @AnotherButler @maciejbocianski what are your thoughts on this?

Reply:
If `ScopedLock` is missing, that is a documentation failure - `ScopedLock` and `ScopedMutexLock` post-date both `CriticalSectionLock` and `DeepSleepLock`, and were an attempt to regularise everything to use `ScopedLock`. (Moving RAII `CriticalSectionLock` to be `ScopedCriticalSectionLock = ScopedLock<CriticalSection>`).

The wind somewhat collapsed in the sails after @sg- blocked the completion of the regularisation on grounds of not wanting to deprecate the just-added-with-little-review `CriticalSectionLock`. (Although in the end we kept the class but deprecated all its existing methods?)

That may explain how the documentation got missed, but I think it's still the latest style agreement. We got the foot in the door with the new style for `ScopedMutexLock`, just didn't manage to change existing ones.

But it means there's no in-tree prototype for what was otherwise agreed to be the model for a singleton scoped lock. That would be `CriticalSectionLock` as it appears in #5420 - thus for this having `MpuXn` and then `ScopedMpuXnLock` using `ScopedLock<MpuXn>`, with a built-in dummy `MpuXn` object.